src/util/locking/MultiThreadedResourceLocker.ts

```diff
@@ -0,0 +1,131 @@
+ * Note that no locks are kept in memory.
+ */
+export class MultiThreadedResourceLocker implements ResourceLocker, Finalizable {
+  protected readonly logger = getLoggerFor(this);
+  private readonly attemptSettings: Required<AttemptSettings>;
+  /** Folder that stores the locks */
+  private readonly lockFolder: string;
+
+  public constructor(rootFilePath: string, attemptSettings: AttemptSettings = {}) {
+    this.attemptSettings = { ...attemptDefaults, ...attemptSettings };
+    this.lockFolder = ensureTrailingSlash(joinFilePath(rootFilePath, ...LOCK_DIR.split('/')));
+    ensureDirSync(this.lockFolder);
+  }
+
+  public async acquire(identifier: ResourceIdentifier): Promise<void> {
+    const { path } = identifier;
+    this.logger.debug(`Acquiring lock for ${path}`);
+    try {
+      const opt = this.generateOptions(identifier, defaultLockOptions);
+      await this.tryFn((): Promise<() => Promise<void>> => lock(path, opt));
+    } catch (err: unknown) {
+      throw new InternalServerError(`Error trying to acquire lock for ${path}. ${err}`);
+    }
+  }
+
+  public async release(identifier: ResourceIdentifier): Promise<void> {
+    const { path } = identifier;
+    this.logger.debug(`Releasing lock for ${path}`);
+
+    try {
+      const opt = this.generateOptions(identifier, defaultUnlockOptions);
+      await this.tryFn((): Promise<void> => unlock(path, opt));
+    } catch (err: unknown) {
+      throw new InternalServerError(`Error trying to release lock for ${path}. ${err}`);
+    }
+  }
+
+  private async tryFn(fn: () => Promise<void | (() => Promise<void>)>): Promise<void> {
+    const { retryCount, retryDelay, retryJitter } = this.attemptSettings;
+    const maxTries = retryCount === -1 ? Number.POSITIVE_INFINITY : retryCount + 1;
+    let tries = 1;
+    let ready = false;
+
+    // Keep going until either you get a lock/release or maxTries has been reached.
+    while (!ready && tries <= maxTries) {
+      try {
+        await fn();
+        ready = true;
+      } catch (err: unknown) {
+        if ((err as any).code === 'ENOTACQUIRED') {
+          throw new InternalServerError('Trying to release unexisting lock.');
+        } else {
+          await wait(retryDelay, retryJitter);
+          tries += 1;
+        }
+      }
+    }
+
+    // Max tries was reached
+    if (tries > maxTries) {
+      const err = `The operation did not succeed after the set maximum of tries (${maxTries}).`;
+      this.logger.warn(err);
+      throw new InternalServerError(err);
+    }
+  }
+
+  /**
+   * Map the identifier path to a unique path inside the {@link lockFolder}.
+   * @param identifier - ResourceIdentifier to generate (Un)LockOptions for.
+   * @returns Full path.
+   */
+  private toLockfilePath(identifier: ResourceIdentifier): string {
+    const hash = createHash('md5');
+    const { path } = identifier;
+    return `${this.lockFolder}/${hash.update(path).digest('hex')}`;
```

Comment:
Due to the `ensureTrailingSlash` above this would actually create paths with double slashes. Would advise just using `joinFilePath` to not have to worry about this either way in general.